### PR TITLE
Random forest parallelization

### PR DESCRIPTION
**Pull request by daraghk**

### CODE REVIEW

**Comment by VolodymyrOrlov** [comment]
@daraghk , thank you very much for your contribution to Smartcore! The code looks good, the only thing that is left are `clippy` warnings and better test coverage. Once you have closed all pending checks we can merge this PR to `development`

**Comment by daraghk** [comment]
@VolodymyrOrlov happy to contribute! I've fixed the code coverage issue but the linting errors are unrelated to any code I touched - you can see the same issues in the other random forest pr which is open at  the moment.

File: src/ensemble/random_forest_classifier.rs [hunk 1]
```diff
@@ -221,41 +240,101 @@ impl<T: RealNumber> RandomForestClassifier<T> {
...
-                criterion: parameters.criterion.clone(),
-                max_depth: parameters.max_depth,
-                min_samples_leaf: parameters.min_samples_leaf,
-                min_samples_split: parameters.min_samples_split,
-            };
-            let tree =
-                DecisionTreeClassifier::fit_weak_learner(x, y, samples, mtry, params, &mut rng)?;
-            trees.push(tree);
-        }
+        let thread_pool = rayon::ThreadPoolBuilder::new()
```
Comment (morenol) [inline src/ensemble/random_forest_classifier.rs:246]: It looks like this fail for WASM targets, Could you explore using this https://github.com/GoogleChromeLabs/wasm-bindgen-rayon? or keep the old behavior for `fit` when using `wasm` target, and this one of non `wasm` targets?.

**Review by morenol — CHANGES_REQUESTED** [review]
Code looks good to me.

Just an issue with WASM targets and multithreading support.

Regarding the clippy issue, I think that a rebase/merge with `development` will fix that.

**Comment by morenol** [comment]
I think that we could do parallelization with a feature flag that is disabled for WASM targets

**Comment by Mec-iS** [comment]
I know but we have no bandwidth to follow up on all these features, creating a feature for one case can be only an increase in complexity with very few benefits. We need to prioritise or allocate more time or have somebody that carries on the task of extending parallelization to more algorithms.

**Pull request closed without merging**